File: Ov2/Account/Account.cpp
#include <iostream> 
#include <stdexcept>
using namespace std; 

class Account{
private:
double balance, interestRate;

public: 
    Account(double balance, double interestRate) : balance(balance), interestRate(interestRate){
        if(balance < 0){
            throw invalid_argument("Balance cannot be negative");
        }
        if(interestRate < 0){
            throw invalid_argument("interest-rate cannot be negative");
        }
    }

    double getBalance(){
        return balance;
    }

    double getInterestRate(){
        return interestRate;
    }

    void setInterestRate(double interestRate){
        if(interestRate < 0){
            throw invalid_argument("interest-rate cannot be negative");
        }
    }

    void deposit(double amount){
        if(amount < 0){
            throw invalid_argument("The amount must be non-negative");
        }
        else{
            balance += amount;
        }
    }

    void withdraw(double amount){
        if((balance - amount) < 0){
            throw invalid_argument("You cannot withdraw more than your balance");
        }
        else{
            balance -= amount;
        }
    }   

    void addInterest(){
        balance += balance *(interestRate / 100);
    }

};

File: src/Classes/Account.go
package Classes

import "fmt"

type Account struct {
	balance      float64
	interestRate float64
}

func NewAccount(initalBalance float64) *Account {
	return &Account{balance: initalBalance}
}

func (a *Account) Deposit(amount float64) {
	a.balance += amount
}

func (a *Account) AddInterest() {
	a.balance += a.balance * a.interestRate
}

func (a *Account) GetBalance() float64 {
	return a.balance
}

func (a *Account) GetInterestRate() float64 {
	return a.interestRate
}

func (a *Account) SetInterestRate(interestRate float64) {
	a.interestRate = interestRate
}

func (a *Account) ToString() string {
	return fmt.Sprintf("Account { balance: %.2f, interestrate: %.2f}", a.balance, a.interestRate)
}
